### PR TITLE
fix(sl-12,#8052): root-cause loop dead-code + import shadow — cycle(train_loader) honors N_ITERS=2000

### PR DESCRIPTION
**Pull request by jsboige**
Grain: DEEP/training-rootcause — lane myia-po-2024:CoursIA-2 — prev: MED/notebook-doc-honesty (c.877 Infer-4 Bayesian-Networks). Coord: autonomous `/continue` cycle (no dispatch this cycle).

## Problem

`SL-12-DifferentiableLogicGateNetworks.ipynb` cell[9] training loop had a **dead-code bug** that capped training at 600 iterations — **never reaching `N_ITERS=2000`**:

```python
for it, (x, y) in enumerate(train_loader):   # DataLoader exhausts after 60000/100 = 600 batches
    ...
    if (it + 1) >= N_ITERS:  # N_ITERS=2000 — DEAD CODE, loop exits at 600 first
        break
```

A `DataLoader` yields exactly `len(dataset)/batch_size = 60000/100 = 600` batches then raises `StopIteration`. So the loop **stopped at iter 600**, and the `break >= N_ITERS` guard was unreachable. **Proven firsthand** by the committed output, which logged `iter 200/2000`, `400/2000`, `600/2000` then `"Entrainement termine en 18.2s"` — the `/2000` display was misleading; the run was 600 iters.

The merged #8446/#8444 were **doc-honesty PRs** that aligned the markdown prose (expected-band, timing) to this committed **600-iter** output. They consecrated the bug's *symptom* rather than fixing its *cause*. This PR fixes the cause.

## Fix (two root-causes, both code-level)

**1. Loop dead-code (cell[9])** — `enumerate(cycle(train_loader))` via `itertools.cycle` restarts the loader indefinitely, so `enumerate` honors `N_ITERS=2000`:

```python
from itertools import cycle
# cycle(train_loader) : un DataLoader s'épuise après len(dataset)/batch_size
# batches (ici 60000/100 = 600), bien avant N_ITERS=2000. Sans cycle(),
# enumerate(train_loader) s'arrête à 600 et le `break >= N_ITERS` est mort.
for it, (x, y) in enumerate(cycle(train_loader)):
```

**2. Import shadow (cell[3])** — bare `import difflogic` resolved to the **site-packages** version (not the SL-12 vendored fork). The site-packages `difflogic.py` has 2 debug `print(self.indices[0].dtype, self.indices[1].dtype)` lines (107, 109) that the vendor fork removed — leaking `torch.int64 torch.int64` ×24000 into cell[9]'s outputs. cell[3] now forces the vendor fork:

```python
import os as _os
_VENDOR = _os.path.join(_os.getcwd(), "vendor", "difflogic")
if _os.path.isdir(_VENDOR):
    sys.path.insert(0, _VENDOR)
import difflogic
```

## C870-L lesson 2 — output-as-truth caught a proxy-file error (self-correction)

My c.870 analysis flagged **both** bugs. This cycle I initially **dropped** the import-shadow fix based on a narrow comparison: `__init__.py` was byte-identical between site-packages and `vendor/difflogic`, so I concluded *"the venv was rebuilt since c.870, the shadow is gone, ship only the loop fix"*. **Wrong.** The first re-exec produced **239 `torch.int64 torch.int64` spam lines** — the shadow was still present.

Investigation: the symptom source is **`difflogic.py`**, not `__init__.py`. `diff vendor/difflogic/.../difflogic.py vs site-packages` shows site-packages has 2 extra debug `print(...)` lines (107, 109). The other 3 module files (`compiled_model.py`, `packbitstensor.py`, `__init__.py`) are all byte-identical — **only `difflogic.py` differs**. So a file-name/proxy comparison (`__init__.py`) missed the real divergence.

**Lesson (C878-L):** re-verify on the **exact source of the symptom**, not a proxy file. The re-exec (output-as-truth) caught what a file-name comparison missed — which is precisely why re-execution is mandatory (C.2) and why a committed output is proof, not decoration (#8052). Applied the cell[3] vendor-forcing fix; second re-exec = **0 spam**.

## Verification (firsthand re-exec)

Papermill end-to-end, kernel `difflogic-sl12`, RTX 3070 Laptop 8GB, `IMPL=python` (CPU kernel impl), MNIST 60k train / 10k test, batch 100:

| Metric | Buggy baseline (committed) | This PR |
|---|---|---|
| Last iter logged | `iter 600/2000` | **`iter 2000/2000`** ✓ |
| Final train loss | (early stop) | **1.134** |
| Train acc | (early stop) | **0.780** |
| Training time | `18.2s` | `57.6s` (3.2× — real 2000 iters) |
| Test acc (cell[10]) | (600-iter model) | **0.830 (8301/10000)** — within documented band (~0.80-0.88 CPU 2K iters) |
| `torch.int64` spam in cell[9] | present | **0** |
| cell[3] resolved module | site-packages | `vendor/difflogic/difflogic/__init__.py` |
| `execution_count` | sequential | **[1..12] sequential, 0 errors** |

SOTA-OK: real difflogic fork (Petersen, NeurIPS 2022, arXiv 2210.08277), real MNIST, real training, real test acc — no workaround. The `~0.83` test acc (vs the paper's ~90%) is honest for CPU + 2K iters; the paper uses 5K iters + GPU (documented in the notebook's expected band).

## Scope

- **1 file** changed: `SL-12-DifferentiableLogicGateNetworks.ipynb` (+252/-159 = 2 source fixes + regenerated training-dependent outputs).
- **Only cell[3] + cell[9] source** changed vs `origin/main` (verified by parsed-source diff, 25 cells, no count drift).
- Regenerated outputs: cell[3] (resolved module path), cell[9] (full 2000-iter training log), cell[10] (test acc 0.830), and downstream training-dependent cells — all consistent (C.2).
- `data/MNIST/` runtime download left **untracked** (not staged).
- `metadata.papermill.output_path` scrubbed to repo-relative (no machine-path leak — secrets-hygiene rule 6 exception).
- CRLF = 0 (`*.ipynb text eol=lf`), JSON valid.
- 0 `raise NotImplementedError` / `assert False` / `1/0` (C.1 clean).

## Context

Supersedes #8446/#8444 **at the code (root-cause) level** — those PRs aligned the markdown to the buggy 600-iter output (doc-honesty); this PR fixes the code cause so the output is genuinely 2000-iter. L898: no open PR on SL-12 (#8446/#8444 merged). [ASK ai-01 SL-12 sequencing] (pending since c.870) is now moot — uncontested, delivered directly.

Not `Closes` — #8052 is the cross-family output-honesty epic spanning many notebooks.

See #8052.

Co-Authored-By: Claude-Code <noreply@anthropic.com>

### CODE REVIEW

**Review by jsboige — COMMENTED** [review]
[Hermes] DEEP review — root-cause fix verified ✓

**Fix 1 (loop dead-code):** `enumerate(cycle(train_loader))` — correct. DataLoader yields `60000/100 = 600` batches then `StopIteration`; the `break >= N_ITERS=2000` was unreachable dead code. `itertools.cycle` is the canonical pattern for indefinite data cycling. Root cause correctly identified.

**Fix 2 (import shadow):** `sys.path.insert(0, vendor/difflogic)` forces the fork over site-packages. The 2 debug `print(self.indices[0].dtype...)` lines in site-packages `difflogic.py` (107, 109) were the spam source. Verified absent in committed output.

**Output authenticity (checked):**
- `iter 2000/2000 | loss 1.134 | train_acc 0.780` ✓ (was `iter 600/2000` dead-stop)
- Training time 57.6s vs 18.2s → ~3.2× ratio matches 2000/600 iters ✓
- Test acc 0.830 within documented band ~0.80-0.88 (CPU 2K iters vs paper ~90% @ 5K+GPU) ✓
- execution_count [1..12] sequential, 0 exceptions ✓
- Timestamps 2026-07-25T19:01-19:02 consistent (single papermill run)

**Minor note:** train_acc oscillates (0.73→0.68→0.75→0.78→...) — normal for difflogic w/ stochastic batching under `cycle()`. Final loss monotonically decreases (1.75→1.13). Not a concern.

No secrets in diff. Cell[3]+cell[9] are the only source changes; rest is regenerated outputs. Solid root-cause PR — supersedes #8446/#8444 at the code level.
